From 01855b819622ce0b14b44a1b0db67edfecabac9d Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Fri, 11 Nov 2022 12:08:38 +0800
Subject: [PATCH 0019/1398] [Huawei] mm,hwpoison: isolate hugetlb_hwpoison_full
 with config HUGETLB_HWPOISON_FULL

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187965

--------------------------------

Isolate hugetlb_hwpoison_full with config HUGETLB_HWPOISON_FULL.

Fixes: cae500b79d40 ("mm: hwpoison: enable memory error handling on 1GB hugepage optionaly")
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig | 1 +
 arch/x86/configs/openeuler_defconfig   | 1 +
 mm/Kconfig                             | 7 +++++++
 mm/memory-failure.c                    | 4 ++++
 4 files changed, 13 insertions(+)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 0fda432ba6c9..13ef0294ab83 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -1067,6 +1067,7 @@ CONFIG_KSM=y
 CONFIG_DEFAULT_MMAP_MIN_ADDR=4096
 CONFIG_ARCH_SUPPORTS_MEMORY_FAILURE=y
 CONFIG_MEMORY_FAILURE=y
+CONFIG_HUGETLB_HWPOISON_FULL=y
 CONFIG_HWPOISON_INJECT=m
 CONFIG_TRANSPARENT_HUGEPAGE=y
 CONFIG_TRANSPARENT_HUGEPAGE_ALWAYS=y
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index 6da13366b192..473635519202 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -1013,6 +1013,7 @@ CONFIG_KSM=y
 CONFIG_DEFAULT_MMAP_MIN_ADDR=4096
 CONFIG_ARCH_SUPPORTS_MEMORY_FAILURE=y
 CONFIG_MEMORY_FAILURE=y
+CONFIG_HUGETLB_HWPOISON_FULL=y
 CONFIG_HWPOISON_INJECT=m
 CONFIG_TRANSPARENT_HUGEPAGE=y
 CONFIG_TRANSPARENT_HUGEPAGE_ALWAYS=y
diff --git a/mm/Kconfig b/mm/Kconfig
index 355a858174c4..e703e1d21dde 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -369,6 +369,13 @@ config MEMORY_FAILURE
 	  even when some of its memory has uncorrected errors. This requires
 	  special hardware support and typically ECC memory.
 
+config HUGETLB_HWPOISON_FULL
+	depends on MEMORY_FAILURE
+	bool "Enable hwpoison for 1G hugetlb"
+	default n
+	help
+	  Enable memory failure for 1G hugetlb.
+
 config HWPOISON_INJECT
 	tristate "HWPoison pages injector"
 	depends on MEMORY_FAILURE && DEBUG_KERNEL && PROC_FS
diff --git a/mm/memory-failure.c b/mm/memory-failure.c
index eefd84154ba6..059b20ed36f8 100644
--- a/mm/memory-failure.c
+++ b/mm/memory-failure.c
@@ -1182,6 +1182,7 @@ static int try_to_split_thp_page(struct page *page, const char *msg)
 	return 0;
 }
 
+#ifdef CONFIG_HUGETLB_HWPOISON_FULL
 static bool hugetlb_hwpoison_full;
 
 static int __init enable_hugetlb_hwpoison_full(char *str)
@@ -1190,6 +1191,9 @@ static int __init enable_hugetlb_hwpoison_full(char *str)
 	return 0;
 }
 early_param("hugetlb_hwpoison_full", enable_hugetlb_hwpoison_full);
+#else
+#define hugetlb_hwpoison_full 0
+#endif
 
 static int memory_failure_hugetlb(unsigned long pfn, int flags)
 {
-- 
2.17.1

